modules/gapi/src/executor/gthreadedexecutor.cpp

```diff
@@ -0,0 +1,525 @@
+            if (exec->allocatesOutputs()) {
+                rmat = exec->allocate(desc);
+            } else {
+                Mat mat;
+                createMat(desc, mat);
+                rmat = make_rmat<RMatOnMat>(mat);
+            }
+        }
+        break;
+
+    case GShape::GSCALAR:
+        if (d.storage == Data::Storage::CONST_VAL)
+        {
+            auto rc = RcDesc{d.rc, d.shape, d.ctor};
+            magazine::bindInArg(m_res, rc, m_gm.metadata(orig_nh).get<ConstValue>().arg);
+        }
+        break;
+
+    case GShape::GARRAY:
+        if (d.storage == Data::Storage::CONST_VAL)
+        {
+            auto rc = RcDesc{d.rc, d.shape, d.ctor};
+            magazine::bindInArg(m_res, rc, m_gm.metadata(orig_nh).get<ConstValue>().arg);
+        }
+        break;
+    case GShape::GOPAQUE:
+        // Constructed on Reset, do nothing here
+        break;
+    case GShape::GFRAME: {
+        // Should be defined by backend, do nothing here
+        break;
+    }
+    default:
+        GAPI_Error("InternalError");
+    }
+}
+
+cv::gimpl::IslandActor::IslandActor(const std::vector<RcDesc>          &in_objects,
+                                    const std::vector<RcDesc>          &out_objects,
+                                    std::shared_ptr<GIslandExecutable> isl_exec,
+                                    Mag                                &res,
+                                    std::mutex                         &m)
+    : m_in_objs(in_objects),
+      m_out_objs(out_objects),
+      m_isl_exec(isl_exec),
+      m_res(res),
+      m_mutex(m) {
+}
+
+void cv::gimpl::IslandActor::run() {
+    cv::gimpl::GThreadedExecutor::Input  i{m_res,  m_in_objs,  m_mutex};
+    cv::gimpl::GThreadedExecutor::Output o{m_res,  m_out_objs, m_mutex};
+    m_isl_exec->run(i, o);
+    // NB: Can't throw an exception there as it's supposed to be
+    // executed in different threads
+    try {
+        o.verify();
+    } catch (...) {
+        m_e = std::current_exception();
+    }
+}
+
+void cv::gimpl::IslandActor::verify() {
+    if (m_e) {
+        std::rethrow_exception(m_e);
+    }
+};
+
+class cv::gimpl::Task {
+public:
+    using F = std::function<void()>;
+    using Ptr = std::shared_ptr<Task>;
+
+    Task(F f, std::vector<Task::Ptr> &&deps);
+
+    void run();
+    void reset() { m_ready_deps.store(0u); }
+
+private:
+    F                     m_f;
+    const uint32_t        m_num_deps;
+    std::atomic<uint32_t> m_ready_deps;
+    std::vector<Task*>    m_dependents;
+};
+
+cv::gimpl::Task::Task(F f, std::vector<Task::Ptr> &&deps)
+    : m_f(f),
+      m_num_deps(static_cast<uint32_t>(deps.size())),
+      m_ready_deps(0u) {
+    for (auto dep : deps) {
+        dep->m_dependents.push_back(this);
+    }
+}
+
+void cv::gimpl::Task::run() {
+    m_f();
+    for (auto* dep : m_dependents) {
+        if (dep->m_ready_deps.fetch_add(1u) == dep->m_num_deps - 1) {
+            cv::gapi::own::ThreadPool::get()->schedule([dep](){
```

Comment:
so the thread pool is shared among the different graphs, right?

Reply:
No, it's not, ThreadPool is available via `get()`` from the `Task` that already in pool.